File: benchmarks/bench_sort.go
package main

import "fmt"

func quicksort(arr []int) []int {
	if len(arr) <= 1 {
		r := make([]int, len(arr))
		copy(r, arr)
		return r
	}
	pivot := arr[0]
	var left, right []int
	for _, x := range arr[1:] {
		if x < pivot {
			left = append(left, x)
		} else {
			right = append(right, x)
		}
	}
	l := quicksort(left)
	r := quicksort(right)
	out := make([]int, 0, len(arr))
	out = append(out, l...)
	out = append(out, pivot)
	out = append(out, r...)
	return out
}

func main() {
	data := make([]int, 0, 1000)
	seed := uint64(42)
	for i := 0; i < 1000; i++ {
		seed = (seed*1103515245 + 12345) % (1 << 31)
		data = append(data, int(seed%10000))
	}
	s := quicksort(data)
	if len(s) != 1000 {
		panic("length")
	}
	for j := 1; j < len(s); j++ {
		if s[j-1] > s[j] {
			panic("order")
		}
	}
	fmt.Println("sorted", len(s), "values")
}

File: benchmarks/bench_sort.py
def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[0]
    left, right = [], []
    for x in arr[1:]:
        (left if x < pivot else right).append(x)
    return quicksort(left) + [pivot] + quicksort(right)

data = []
seed = 42
for _ in range(1000):
    seed = (seed * 1103515245 + 12345) % (2 ** 31)
    data.append(seed % 10000)

sorted_data = quicksort(data)
assert len(sorted_data) == 1000
for j in range(1, 1000):
    assert sorted_data[j - 1] <= sorted_data[j]

print(f"sorted {len(sorted_data)} values")
